src/data/nat/factorization/basic.lean

```diff
@@ -382,6 +384,20 @@ begin
     simp [←factorization_le_iff_dvd he_pos hd_pos, h1, hea', heb'] },
 end
 
+lemma factorization_eq_card_pow_dvd {n p b : ℕ} (pp : p.prime) (hn : 0 < n) (hb : log p n < b) :
```

Comment:
It seems that the standard way to write a positivity hypothesis for nat is `hn : n ≠ 0`.

I think it would be good to develop some properties of this set of exponents -- and a nice benefit is that you don't need to supply `b` to use the lemma. Here's an attempt :

```lean
lemma set_of_pow_dvd_eq {n p b : ℕ} (pp : p.prime) (hn : n ≠ 0) (hb : log p n < b) :
  {i : ℕ | i ≠ 0 ∧ p ^ i ∣ n} = (finset.Ico 1 b).filter (λ i, p ^ i ∣ n) :=
begin
  ext i,
  simp only [pp.pow_dvd_iff_le_factorization hn, one_le_iff_ne_zero, set.mem_set_of_eq, coe_filter,
    set.mem_sep_eq, ne.def, coe_Ico, set.mem_Ico, and.congr_left_iff, iff_self_and],
  intros h _,
  refine lt_of_le_of_lt _ hb,
  rw ← pow_le_iff_le_log pp.one_lt (pos_iff_ne_zero.mpr hn),
  exact le_trans (pow_le_pow pp.one_lt.le h) (pow_factorization_le p hn),
end

def set_of_pow_dvd.fintype {n p : ℕ} (pp : p.prime) (hn : n ≠ 0) :
  fintype {i : ℕ | i ≠ 0 ∧ p ^ i ∣ n} :=
fintype.of_finset ((finset.Ico 1 (nat.find (exists_gt (log p n)))).filter (λ i, p ^ i ∣ n)) $ λ i,
begin
  rw set_of_pow_dvd_eq pp hn (nat.find_spec (exists_gt (log p n))),
  refl,
end

lemma set_of_pow_dvd_subset {n p b : ℕ} (pp : p.prime) (hn : n ≠ 0) (hb : log p n < b) :
  {i : ℕ | i ≠ 0 ∧ p ^ i ∣ n} ⊆ set.Ico 1 b :=
begin
  rw [set_of_pow_dvd_eq pp hn hb, ← set.to_finset_mono],
  intro i,
  simp { contextual := tt },
end

lemma factorization_eq_card_pow_dvd {n p : ℕ} (pp : p.prime) (hn : n ≠ 0) :
  by haveI := set_of_pow_dvd.fintype pp hn; exact -- maybe make it take a `fintype` instance?
  n.factorization p = ({i : ℕ | i ≠ 0 ∧ p ^ i ∣ n}).to_finset.card :=
begin
  rw [←nat.add_sub_cancel (n.factorization p) 1, ←card_Ico],
  apply congr_arg card,
  ext,
  simp [lt_succ_iff, one_le_iff_ne_zero, pp.pow_dvd_iff_le_factorization hn],
end
```

Reply:
> It seems that the standard way to write a positivity hypothesis for nat is `hn : n ≠ 0`.

Yes, you're right, I failed to fix this from `multiplicity_eq_card_pow_dvd`.